### DISCUSSION TITLE
Баги в работе tree

### DISCUSSION THREAD

**Discussion by omqmo** [taiga-family/taiga-ui#5594] [Q&A] [answered]
Динамическое создание дерева. Создаю первый уровень - элемент отображается. Прихожу к выводу, что с моей стороны все сделано правильно ( я имею ввиду стратегия onpush и все такое). Но если создать вложенный элемент, то он не рендирится и у родителя даже не создается expand. Если перезагрузить, то все работает. И независомо от уровня только-что созданные соседнии узлы всегда отображаются, а вот дочернии нет. И причина в том, что родитель не создает expand. И начинает работать только если переключить тугл у предка, не у родителя, а именно у предка.

**Comment by omqmo** [comment]
`ChangeDetectorRef`, даже с задержкой по setTimeout, не помогает...

**Comment by splincode** [comment]
Добрый день, не могли бы вы добавить ссылку на stackblitz с воспроизведением?

**Reply by omqmo** [reply]
stackblitz, это моя ахиллесова пята. [Вот ссылка на репо](https://github.com/omqmo/tree-examples), а вот [на stackblitz](https://stackblitz.com/github/omqmo/tree-examples), который, почему-то, не работает. Вот почему он не работает? Везде говорят, что просто дай ссылку на репо и все. Но при этому ошибка 

> Cannot resolve type entity i9.PolymorpheusModule to symbol

По примеру - нажимаете кнопку, ничего не создается, что иллюстрирует проблему.

**Reply by splincode** [reply]
воспользуйтесь в качестве шаблона
https://taiga-ui.dev/stackblitz

**Reply by omqmo** [reply]
> Module '"@angular/core"' has no exported member 'signal'.

**Reply by omqmo** [reply]
Все равно выяснять причину на stackblitz не сможете. Избавьте меня хотя бы багов связанных с ним :) Мне вообще уже кажется, что я бесплатно на вас тестировщиком работаю.

**Reply by omqmo** [reply]
И вы конечно разберетесь, но все равно стоит сказать, что не работает вот так
```
tree.children[ 0 ].children.push( node );
```

Если добавлять чилдов самому корню, то работает
```
tree.children[ 0 ].push( node );
```

**Reply by splincode** [reply]
`Module '"@angular/core"' has no exported member 'signal'.`
Переделайте пожалуйста пример без использовать signal

**Reply by omqmo** [reply]
Хорошо, а вы пока подумайте над тем как-бы сделать PolymorpheusModule устанавливаемым.

**Reply by waterplea** [reply]
> И вы конечно разберетесь, но все равно стоит сказать, что не работает вот так
> ```
> tree.children[ 0 ].children.push( node );
> ```
> 
> Если добавлять чилдов самому корню, то работает
> ```
> tree.children[ 0 ].push( node );
> ```
> 

Так ты мутируешь массив, OnPush такое не подхватывает же. Пока не смотрел чего там, просто глаз зацепился, может и не про то совсем.

**Reply by omqmo** [reply]
Ещё один вопрос - tui-tree в качестве value ожидает ТОЛЬКО иммутабельное значение? Другими словами, должены ли применяться изменения если они произошли в родители мутабельно, то есть без пересоздания самого узла?

**Reply by omqmo** [reply]
@waterplea вот я об этом уже тоже подумал. value обязательно должна быть иммутабельной  структурой?

**Reply by waterplea** [reply]
> @waterplea вот я об этом уже тоже подумал. value обязательно должна быть иммутабельной  структурой? 

Да, там не только OnPush, но и distinctUntilChanged есть, так что попробуй переписать на filter/concat, вместо splice/push. Скорее всего поможет.

**Reply by omqmo** [reply]
@waterplea понятно. Тогда вопрос снимается.

**Comment by waterplea** [comment, marked as answer]
Проблема была в мутировании данных в дочерних нодах, что не поддерживается компонентом.

**Comment by omqmo** [comment]
@waterplea еще один вопрос по теме. Если есть цепочка a > b > c, то при изменении последнего ребенка нужно пересоздать всю цепочку или будет достаточно только c и a?

**Reply by waterplea** [reply]
Подозреваю, что должно хватить обновить только тот массив, в котором дети поменялись.
